File: school-ms/frontend/src/services/courseService.ts
import api from './api';

export interface Course {
  id?: number;
  name: string;
  department: string;
  teacherId: number;
  credits: number;
  capacity: number;
  enrolled: number;
}

export const courseService = {
  getAll: () => api.get<Course[]>('courses'),  // Removed /api prefix to avoid duplication
  getById: (id: number) => api.get<Course>(`courses/${id}`),  // Removed /api prefix
  create: (course: Course) => api.post<Course>('courses', course),  // Removed /api prefix
  update: (id: number, course: Course) => api.put<Course>(`courses/${id}`, course),  // Removed /api prefix
  delete: (id: number) => api.delete(`courses/${id}`),  // Removed /api prefix
  enrollStudent: (courseId: number, studentId: number) => 
    api.post<void>(`courses/${courseId}/enroll`, { studentId }),  // Removed /api prefix
  unenrollStudent: (courseId: number, studentId: number) => 
    api.delete<void>(`courses/${courseId}/enroll?studentId=${studentId}`),  // Removed /api prefix
  getAllCourses: () => api.get<Course[]>('courses'),  // Alias for getAll for compatibility
};

File: school-ms/backend/school-app/src/main/java/com/school/course/service/CourseService.java
package com.school.course.service;

import com.school.course.model.Course;
import java.util.List;

public interface CourseService {
    List<Course> getAllCourses();

    Course getCourseById(Long id);

    Course createCourse(Course course);

    Course updateCourse(Long id, Course course);

    void deleteCourse(Long id);

    List<Course> getCoursesByDepartment(String department);

    List<Course> getCoursesByTeacherId(Long teacherId);

    void enrollStudent(Long courseId, Long studentId);

    void unenrollStudent(Long courseId, Long studentId);
}